src/ecTyping.ml

```diff
@@ -2249,6 +2258,229 @@ and transmod_body ~attop (env : EcEnv.env) x params (me:pmodule_expr) =
+          if after then
+            i :: (s_subst ts s).s_node
+          else
+            (s_subst ts s).s_node @ [i]
+        | Pups_del -> []
+      in
+
+      (* Semantics for condition updating *)
+      (* `i` is the target of the update, and `tl` is the instr suffix. *)
+      let eval_cupdate cp_loc env cup i tl =
+        match cup with
+        (* Insert an if with condition `e` with body `tl` *)
+        | Pupc_add e ->
+          let loc = e.pl_loc in
+          let ue  = UE.create (Some []) in
+          let e, ty = transexp env `InProc ue e in
+          let ts = Tuni.subst (UE.close ue) in
+          let ty = ty_subst ts ty in
+          unify_or_fail env ue loc ~expct:tbool ty;
+          [i] @ [i_if (e_subst ts e, stmt tl, s_empty)]
+
+        (* Change the condition expression to `e` for a conditional instr `i` *)
+        | Pupc_mod e -> begin
+          let loc = e.pl_loc in
+          let ue  = UE.create (Some []) in
+          let e, ty = transexp env `InProc ue e in
+          let ts = Tuni.subst (UE.close ue) in
+          let ty = ty_subst ts ty in
+          match i.i_node with
+          | Sif (_, t, f) -> 
+            unify_or_fail env ue loc ~expct:tbool ty;
+            [i_if (e_subst ts e, t, f)] @ tl
+          | Smatch (p, bs) ->
+            unify_or_fail env ue loc ~expct:p.e_ty ty;
+            [i_match (e_subst ts e, bs)] @ tl
+          | Swhile (_, t) ->
+            unify_or_fail env ue loc ~expct:tbool ty;
+            [i_while (e_subst ts e, t)] @ tl
+          | _ ->
+            tyerror cp_loc env (InvalidModUpdate MUE_InvalidTargetCond);
+          end
+
+        (* Collapse a conditional `i` to a specific branch `bs` *)
+        | Pupc_del bs -> begin
+          let bs = trans_codepos_brsel bs in
+          match i.i_node, bs with
+          | Sif (_, t, _), `Cond true -> t.s_node
+          | Sif (_, _, f), `Cond false -> f.s_node
+          | Swhile (_, t), `Cond true -> t.s_node
+          | Smatch (_e, _bs), `Match _cn ->
+            (* FIXME: need to introduce binding for constructor pattern variables *)
+            (* match e with | C a b c => b | ...  ---> (a, b, c) <- oget (get_as_C e); b *)
+            (*
+            let _, indt, _ = oget (EcEnv.Ty.get_top_decl e.e_ty env) in
+            let indt = oget (EcDecl.tydecl_as_datatype indt) in
+            let cnames = List.fst indt.tydt_ctors in
+            let ix, _ = List.findi (fun _ n -> EcSymbols.sym_equal cn n) cnames in
+            let _, r = List.split_at ix bs in
+            let _p, b, _ = match r with (p, b) :: r -> p, b, r | _ -> assert false in
+            b.s_node @ tl
+            *)
+            tyerror cp_loc env (InvalidModUpdate MUE_InvalidTargetCond);
+          | _ ->
+            tyerror cp_loc env (InvalidModUpdate MUE_InvalidTargetCond);
+          end
+      in
+
+      (* Apply each of updates in reverse *)
+      (* NOTE: This is with the expectation that the user entered them in chronological order. *)
```

Comment:
Don't we want to check that?

Reply:
I don't think we can—this is user intent we'd need to check. (Chronological here is "Do this transformation, then that transformation.") Is that correct, @Cameron-Low?